From 5725182d765dfaa5a7dc6b2e2cfe590270e34011 Mon Sep 17 00:00:00 2001
From: Derek Watson <derekw1996@gmail.com>
Date: Tue, 25 Oct 2022 23:50:15 -0400
Subject: [PATCH] input: misc: add TWL6030 support to twl_pwrbutton

Support the TWL6030 power button in the now-generic twl_pwrbutton driver.

Based on [1] but applied on top of [2].

[1] https://github.com/klokik/jem_linux_kernel/commit/226f44376d8b5f9a5a0c3e2616c72f308771d16e
[2] https://patches.linaro.org/project/linux-pm/list/?series=184222
---
 drivers/input/misc/twl-pwrbutton.c | 32 ++++++++++++++++++++++++++++--
 include/linux/mfd/twl.h            |  1 +
 2 files changed, 31 insertions(+), 2 deletions(-)

diff --git a/drivers/input/misc/twl-pwrbutton.c b/drivers/input/misc/twl-pwrbutton.c
index 5614c5003828..4ff9d93b874a 100644
--- a/drivers/input/misc/twl-pwrbutton.c
+++ b/drivers/input/misc/twl-pwrbutton.c
@@ -31,7 +31,8 @@
 
 #define PWR_PWRON_IRQ (1 << 0)
 
-#define STS_HW_CONDITIONS 0xf
+#define TWL4030_STS_HW_CONDITIONS 0x0f
+#define TWL6030_STS_HW_CONDITIONS 0x21
 
 static irqreturn_t powerbutton_irq(int irq, void *_pwr)
 {
@@ -39,7 +40,13 @@ static irqreturn_t powerbutton_irq(int irq, void *_pwr)
 	int err;
 	u8 value;
 
-	err = twl_i2c_read_u8(TWL_MODULE_PM_MASTER, &value, STS_HW_CONDITIONS);
+	if (twl_class_is_6030())
+		err = twl_i2c_read_u8(TWL6030_MODULE_ID0, &value,
+				TWL6030_STS_HW_CONDITIONS);
+	else
+		err = twl_i2c_read_u8(TWL_MODULE_PM_MASTER, &value,
+				TWL4030_STS_HW_CONDITIONS);
+
 	if (!err)  {
 		pm_wakeup_event(pwr->dev.parent, 0);
 		input_report_key(pwr, KEY_POWER, value & PWR_PWRON_IRQ);
@@ -69,6 +76,13 @@ static int twl_pwrbutton_probe(struct platform_device *pdev)
 	pwr->phys = "twl_pwrbutton/input0";
 	pwr->dev.parent = &pdev->dev;
 
+	if (twl_class_is_6030()) {
+		twl6030_interrupt_unmask(TWL6030_PWRON_INT_MASK,
+			REG_INT_MSK_LINE_A);
+		twl6030_interrupt_unmask(TWL6030_PWRON_INT_MASK,
+			REG_INT_MSK_STS_A);
+	}
+
 	err = devm_request_threaded_irq(&pdev->dev, irq, NULL, powerbutton_irq,
 			IRQF_TRIGGER_FALLING | IRQF_TRIGGER_RISING |
 			IRQF_ONESHOT,
@@ -89,9 +103,22 @@ static int twl_pwrbutton_probe(struct platform_device *pdev)
 	return 0;
 }
 
+static int twl_pwrbutton_remove(struct platform_device *pdev)
+{
+	if (twl_class_is_6030()) {
+		twl6030_interrupt_mask(TWL6030_PWRON_INT_MASK,
+			REG_INT_MSK_LINE_A);
+		twl6030_interrupt_mask(TWL6030_PWRON_INT_MASK,
+			REG_INT_MSK_STS_A);
+	}
+
+	return 0;
+}
+
 #ifdef CONFIG_OF
 static const struct of_device_id twl_pwrbutton_dt_match_table[] = {
 	{ .compatible = "ti,twl4030-pwrbutton" },
+	{ .compatible = "ti,twl6030-pwrbutton" },
 	{},
 };
 MODULE_DEVICE_TABLE(of, twl_pwrbutton_dt_match_table);
@@ -99,6 +126,7 @@ MODULE_DEVICE_TABLE(of, twl_pwrbutton_dt_match_table);
 
 static struct platform_driver twl_pwrbutton_driver = {
 	.probe		= twl_pwrbutton_probe,
+	.remove		= twl_pwrbutton_remove,
 	.driver		= {
 		.name	= "twl_pwrbutton",
 		.of_match_table = of_match_ptr(twl_pwrbutton_dt_match_table),
diff --git a/include/linux/mfd/twl.h b/include/linux/mfd/twl.h
index 089e8942223a..ec2ea1f817c6 100644
--- a/include/linux/mfd/twl.h
+++ b/include/linux/mfd/twl.h
@@ -113,6 +113,7 @@ enum twl6030_module_ids {
 #define REG_INT_MSK_STS_C		0x08
 
 /* MASK INT REG GROUP A */
+#define TWL6030_PWRON_INT_MASK		0x01
 #define TWL6030_PWR_INT_MASK 		0x07
 #define TWL6030_RTC_INT_MASK 		0x18
 #define TWL6030_HOTDIE_INT_MASK 	0x20
-- 
2.30.2

